File: u.js
// The U device. Universal device, unknown device, U. Any U name reaches it.
//
// Three faces: 1, 0 and u. It combines them without ever turning u into a
// guess, and it will not hand you a u that does not say what would settle it.
//
// The fingerprint at the bottom is computed from the four functions above it.
// It is not read from a comment. Delete this header and it comes out the same.
// chuumind.com/u says whose it is.

import { pathToFileURL } from "node:url";

export const u = why => {
  if (!why) throw new Error("a u names the one observation that would settle it");
  return "u";
};

export const not = a => (a === "u" ? "u" : a === "1" ? "0" : "1");
export const and = (a, b) => (a === "0" || b === "0" ? "0" : a === "u" || b === "u" ? "u" : "1");
export const or  = (a, b) => (a === "1" || b === "1" ? "1" : a === "u" || b === "u" ? "u" : "0");

// Every face the device can show, in the order he says them.
export const faces = ["1", "0", "u"];

// The tape is the device's whole behaviour written down: not over each face,
// and over each pair, or over each pair, then one character for whether the
// device still refuses a bare u. 22 characters. This is the fingerprint.
export const tape = () =>
  faces.map(not).join("") +
  faces.map(a => faces.map(b => and(a, b)).join("")).join("") +
  faces.map(a => faces.map(b => or(a, b)).join("")).join("") +
  (() => { try { u(); return "0"; } catch { return "1"; } })();

export const KNOWN = "01u10u000u0u11110u1uu1";

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  const t = tape();
  console.log("tape      ", t);
  console.log("known     ", KNOWN);
  console.log("match     ", t === KNOWN ? "1" : "0");
  console.log("all is    ", or("u", not("u")));   // not 1
  console.log("and isn't ", and("u", not("u")));  // not 0
}

File: u.py
"""The U device. Universal device, unknown device, U. Any U name reaches it.

Three faces: 1, 0 and u. It combines them without ever turning u into a guess,
and it will not hand you a u that does not say what would settle it.

The fingerprint at the bottom is computed from the four functions above it.
It is not read from a docstring. Delete this header and it comes out the same.
chuumind.com/u says whose it is.
"""


def u(why=None):
    if not why:
        raise ValueError("a u names the one observation that would settle it")
    return "u"


def not_(a):
    return "u" if a == "u" else "0" if a == "1" else "1"


def and_(a, b):
    if a == "0" or b == "0":
        return "0"
    return "u" if a == "u" or b == "u" else "1"


def or_(a, b):
    if a == "1" or b == "1":
        return "1"
    return "u" if a == "u" or b == "u" else "0"


# Every face the device can show, in the order he says them.
FACES = ("1", "0", "u")


def guard():
    try:
        u()
    except ValueError:
        return "1"
    return "0"


# The tape is the device's whole behaviour written down: not over each face,
# and over each pair, or over each pair, then one character for whether the
# device still refuses a bare u. 22 characters. This is the fingerprint.
def tape():
    return (
        "".join(not_(a) for a in FACES)
        + "".join(and_(a, b) for a in FACES for b in FACES)
        + "".join(or_(a, b) for a in FACES for b in FACES)
        + guard()
    )


KNOWN = "01u10u000u0u11110u1uu1"

if __name__ == "__main__":
    t = tape()
    print("tape      ", t)
    print("known     ", KNOWN)
    print("match     ", "1" if t == KNOWN else "0")
    print("all is    ", or_("u", not_("u")))   # not 1
    print("and isn't ", and_("u", not_("u")))  # not 0
